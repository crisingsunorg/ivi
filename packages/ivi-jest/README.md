`ivi-jest` provides a collection of useful tools for testing with [Jest](https://jestjs.io/).

## Motivation

Jest resets global state for every test module, but all tests in a test module share the same global state. This package
provides a set of tools to make it easier to write isolated unit tests.

## API

### DOM

`useResetDOMGlobalEventListeners()` tracks event handlers for a window and document objects and removes them after
each test.

`useResetDOM()` enables following resets:

- `useResetDOMGlobalEventListeners()`

`useDOMElement()` creates a DOM element and optionally mounts it to the document body.

`useRequestAnimationFrame()` mocks `requestAnimationFrame()` and `cancelAnimationFrame()`.

### Modules

`useResetModules()` resets modules before each test.

`useModule()` imports module before each test.

### Mock

`useMockFn()` creates a mock function and resets it after each test.

`useSpyOn()` creates a spy object and resets it after each test.

### Environment

`useEnv()` overrides `process.env` before each tests and restores to the original value after each test.

### Promise

`usePromiseQueue()` adds support for waiting until promise tasks are executed.

### Timers

`useTimers()` mocks timers.

### Computed Value

`useComputedValue()` creates a value that will be recomputed before each test.
