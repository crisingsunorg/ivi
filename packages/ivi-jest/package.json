{
  "private": true,
  "name": "ivi-jest",
  "version": "0.0.1",
  "main": "dist/main/index.js",
  "module": "dist/module/index.js",
  "sideEffects": false,
  "files": [
    "dist",
    "src",
    "!dist/**/*.tsbuildinfo",
    "!src/**/__tests__",
    "README.md",
    "LICENSE"
  ],
  "scripts": {
    "prepublishOnly": "npm run clean && npm run dist",
    "clean": "rm -rf ./dist",
    "dist": "npm run dist:main && npm run dist:module",
    "dist:main": "../../node_modules/.bin/tsc -p ./tsconfig.build.main.json",
    "dist:module": "../../node_modules/.bin/tsc -p ./tsconfig.build.module.json"
  },
  "dependencies": {
    "jsdom": "^14.0.0"
  },
  "peerDependencies": {},
  "devDependencies": {},
  "description": "Jest utils for ivi library",
  "license": "MIT",
  "author": {
    "name": "Boris Kaul",
    "email": "localvoid@gmail.com",
    "url": "https://github.com/localvoid"
  },
  "keywords": [
    "ivi"
  ],
  "homepage": "https://github.com/localvoid/ivi",
  "repository": {
    "type": "git",
    "url": "https://github.com/localvoid/ivi"
  },
  "bugs": {
    "url": "https://github.com/localvoid/ivi/issues"
  },
  "github": "https://github.com/localvoid/ivi"
}
