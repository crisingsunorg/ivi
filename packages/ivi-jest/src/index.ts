export { useComputedValue } from "./value";
export { useEnv } from "./env";
export { useMockFn, useSpyOn } from "./mock";
export { useResetModules, useModule, useIVI, useHTML, useSVG, useTest } from "./module";
export { useResetDOMGlobalEventListeners, useResetDOM, useDOMElement } from "./dom";
