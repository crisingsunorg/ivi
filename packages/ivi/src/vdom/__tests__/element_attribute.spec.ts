import { useResetJSDOM, useResetModules, useSpyOn, useDOMElement, useHTML, useTest } from "ivi-jest";

useResetJSDOM();
useResetModules();
const root = useDOMElement();
const h = useHTML();
const t = useTest();
const _ = void 0;
const r = (attrs?: {}) => t.render<HTMLElement>(h.div(_, attrs), root()).domNode!;

describe("element attribute", () => {
  describe("mount", () => {
    const setAttribute = useSpyOn(() => Element.prototype, "setAttribute");
    const removeAttribute = useSpyOn(() => Element.prototype, "removeAttribute");

    test("undefined", () => {
      expect(r()).toMatchSnapshot();
      expect(setAttribute().mock.calls).toEqual([]);
      expect(removeAttribute().mock.calls).toEqual([]);
    });

    test("empty attributes object", () => {
      expect(r({})).toMatchSnapshot();
      expect(setAttribute().mock.calls).toEqual([]);
      expect(removeAttribute().mock.calls).toEqual([]);
    });

    test("attribute with undefined value", () => {
      expect(r({ width: void 0 })).toMatchSnapshot();
      expect(setAttribute().mock.calls).toEqual([]);
      expect(removeAttribute().mock.calls).toEqual([]);
    });

    test("attribute with empty string value", () => {
      expect(r({ width: "" })).toMatchSnapshot();
      expect(setAttribute().mock.calls).toEqual([["width", ""]]);
      expect(removeAttribute().mock.calls).toEqual([]);
    });

    test("attribute with string value", () => {
      expect(r({ width: "10px" })).toMatchSnapshot();
      expect(setAttribute().mock.calls).toEqual([["width", "10px"]]);
      expect(removeAttribute().mock.calls).toEqual([]);
    });

    test("two attributes with string values", () => {
      expect(r({ width: "10px", height: "20px" })).toMatchSnapshot();
      expect(setAttribute().mock.calls).toEqual([["width", "10px"], ["height", "20px"]]);
      expect(removeAttribute().mock.calls).toEqual([]);
    });

    test("attribute with true value", () => {
      expect(r({ boolean: true })).toMatchSnapshot();
      expect(setAttribute().mock.calls).toEqual([["boolean", ""]]);
      expect(removeAttribute().mock.calls).toEqual([]);
    });

    test("attribute with false value", () => {
      expect(r({ boolean: false })).toMatchSnapshot();
      expect(setAttribute().mock.calls).toEqual([]);
      expect(removeAttribute().mock.calls).toEqual([]);
    });
  });

  describe("update", () => {
    describe("undefined to", () => {
      beforeEach(() => { r(); });
      const setAttribute = useSpyOn(() => Element.prototype, "setAttribute");
      const removeAttribute = useSpyOn(() => Element.prototype, "removeAttribute");

      test("undefined", () => {
        expect(r()).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("empty attributes object", () => {
        expect(r({})).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with undefined value", () => {
        expect(r({ width: void 0 })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with empty string value", () => {
        expect(r({ width: "" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", ""]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with string value", () => {
        expect(r({ width: "10px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "10px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with true value", () => {
        expect(r({ boolean: true })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["boolean", ""]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with false value", () => {
        expect(r({ boolean: false })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("two attributes with string values", () => {
        expect(r({ width: "10px", height: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "10px"], ["height", "20px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });
    });

    describe("attribute with undefined value to", () => {
      beforeEach(() => { r({ width: void 0 }); });
      const setAttribute = useSpyOn(() => Element.prototype, "setAttribute");
      const removeAttribute = useSpyOn(() => Element.prototype, "removeAttribute");

      test("undefined", () => {
        expect(r()).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("empty attribute object", () => {
        expect(r({})).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with undefined value", () => {
        expect(r({ width: void 0 })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with empty string value", () => {
        expect(r({ width: "" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", ""]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with string value", () => {
        expect(r({ width: "10px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "10px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with true value", () => {
        expect(r({ boolean: true })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["boolean", ""]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with false value", () => {
        expect(r({ boolean: false })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("one attribute with same value and one with different", () => {
        expect(r({ width: void 0, height: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["height", "20px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("two attributes with different values", () => {
        expect(r({ width: "30px", height: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "30px"], ["height", "20px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("two different attributes", () => {
        expect(r({ left: "30px", height: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["left", "30px"], ["height", "20px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });
    });

    describe("attribute with empty string value to", () => {
      beforeEach(() => { r({ width: "" }); });
      const setAttribute = useSpyOn(() => Element.prototype, "setAttribute");
      const removeAttribute = useSpyOn(() => Element.prototype, "removeAttribute");

      test("undefined", () => {
        expect(r()).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });

      test("empty attribute object", () => {
        expect(r({})).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });

      test("attribute with undefined value", () => {
        expect(r({ width: void 0 })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });

      test("attribute with empty string value", () => {
        expect(r({ width: "" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with different value", () => {
        expect(r({ width: "10px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "10px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with true value", () => {
        expect(r({ width: true })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", ""]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with false value", () => {
        expect(r({ width: false })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });

      test("one attribute with same value and one with different", () => {
        expect(r({ width: "", height: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["height", "20px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("two attributes with different values", () => {
        expect(r({ width: "30px", height: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "30px"], ["height", "20px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("two different attributes", () => {
        expect(r({ left: "30px", height: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["left", "30px"], ["height", "20px"]]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });
    });

    describe("attribute with string value to", () => {
      beforeEach(() => { r({ width: "10px" }); });
      const setAttribute = useSpyOn(() => Element.prototype, "setAttribute");
      const removeAttribute = useSpyOn(() => Element.prototype, "removeAttribute");

      test("undefined", () => {
        expect(r()).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });

      test("empty attribute object", () => {
        expect(r({})).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });

      test("attribute with undefined value", () => {
        expect(r({ width: void 0 })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });

      test("attribute with empty string value", () => {
        expect(r({ width: "" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", ""]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with same value", () => {
        expect(r({ width: "10px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with different value", () => {
        expect(r({ width: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "20px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with true value", () => {
        expect(r({ width: true })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", ""]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with false value", () => {
        expect(r({ width: false })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });

      test("one attribute with same value and one with different", () => {
        expect(r({ width: "10px", height: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["height", "20px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("two attributes with different values", () => {
        expect(r({ width: "30px", height: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "30px"], ["height", "20px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("two different attributes", () => {
        expect(r({ left: "30px", height: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["left", "30px"], ["height", "20px"]]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });
    });

    describe("attribute with false value to", () => {
      beforeEach(() => { r({ width: false }); });
      const setAttribute = useSpyOn(() => Element.prototype, "setAttribute");
      const removeAttribute = useSpyOn(() => Element.prototype, "removeAttribute");

      test("undefined", () => {
        expect(r()).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("empty attribute object", () => {
        expect(r({})).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with undefined value", () => {
        expect(r({ width: void 0 })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with empty string value", () => {
        expect(r({ width: "" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", ""]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with string value", () => {
        expect(r({ width: "10px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "10px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with true value", () => {
        expect(r({ width: true })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", ""]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with false value", () => {
        expect(r({ width: false })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });
    });

    describe("attribute with true value to", () => {
      beforeEach(() => { r({ width: true }); });
      const setAttribute = useSpyOn(() => Element.prototype, "setAttribute");
      const removeAttribute = useSpyOn(() => Element.prototype, "removeAttribute");

      test("undefined", () => {
        expect(r()).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });

      test("empty attribute object", () => {
        expect(r({})).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });

      test("attribute with undefined value", () => {
        expect(r({ width: void 0 })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });

      test("attribute with empty string value", () => {
        expect(r({ width: "" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", ""]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with string value", () => {
        expect(r({ width: "10px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "10px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with true value", () => {
        expect(r({ width: true })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("attribute with false value", () => {
        expect(r({ width: false })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"]]);
      });
    });

    describe("two attributes with string values to", () => {
      beforeEach(() => { r({ width: "10px", height: "20px" }); });
      const setAttribute = useSpyOn(() => Element.prototype, "setAttribute");
      const removeAttribute = useSpyOn(() => Element.prototype, "removeAttribute");

      test("undefined", () => {
        expect(r()).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"], ["height"]]);
      });

      test("empty attribute object", () => {
        expect(r({})).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"], ["height"]]);
      });

      test("attribute with undefined value", () => {
        expect(r({ width: void 0 })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([["width"], ["height"]]);
      });

      test("attribute with empty string value", () => {
        expect(r({ width: "" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", ""]]);
        expect(removeAttribute().mock.calls).toEqual([["height"]]);
      });

      test("one attribute with different value", () => {
        expect(r({ width: "30px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "30px"]]);
        expect(removeAttribute().mock.calls).toEqual([["height"]]);
      });

      test("two attributes with same values", () => {
        expect(r({ width: "10px", height: "20px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("two attributes with different values", () => {
        expect(r({ width: "30px", height: "40px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["width", "30px"], ["height", "40px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("one attribute with same value and one with different", () => {
        expect(r({ width: "10px", height: "30px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["height", "30px"]]);
        expect(removeAttribute().mock.calls).toEqual([]);
      });

      test("two different attributes", () => {
        expect(r({ left: "30px", right: "40px" })).toMatchSnapshot();
        expect(setAttribute().mock.calls).toEqual([["left", "30px"], ["right", "40px"]]);
        expect(removeAttribute().mock.calls).toEqual([["width"], ["height"]]);
      });
    });
  });
});
