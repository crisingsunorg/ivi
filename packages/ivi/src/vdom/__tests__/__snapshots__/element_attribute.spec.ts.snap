// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`element attribute mount attribute with empty string value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute mount attribute with false value 1`] = `<div />`;

exports[`element attribute mount attribute with string value 1`] = `
<div
  width="10px"
/>
`;

exports[`element attribute mount attribute with true value 1`] = `
<div
  boolean=""
/>
`;

exports[`element attribute mount attribute with undefined value 1`] = `<div />`;

exports[`element attribute mount empty attributes object 1`] = `<div />`;

exports[`element attribute mount two attributes with string values 1`] = `
<div
  height="20px"
  width="10px"
/>
`;

exports[`element attribute mount undefined 1`] = `<div />`;

exports[`element attribute update attribute with empty string value to attribute with different value 1`] = `
<div
  width="10px"
/>
`;

exports[`element attribute update attribute with empty string value to attribute with empty string value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute update attribute with empty string value to attribute with false value 1`] = `<div />`;

exports[`element attribute update attribute with empty string value to attribute with true value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute update attribute with empty string value to attribute with undefined value 1`] = `<div />`;

exports[`element attribute update attribute with empty string value to empty attribute object 1`] = `<div />`;

exports[`element attribute update attribute with empty string value to one attribute with same value and one with different 1`] = `
<div
  height="20px"
  width=""
/>
`;

exports[`element attribute update attribute with empty string value to two attributes with different values 1`] = `
<div
  height="20px"
  width="30px"
/>
`;

exports[`element attribute update attribute with empty string value to two different attributes 1`] = `
<div
  height="20px"
  left="30px"
/>
`;

exports[`element attribute update attribute with empty string value to undefined 1`] = `<div />`;

exports[`element attribute update attribute with false value to attribute with empty string value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute update attribute with false value to attribute with false value 1`] = `<div />`;

exports[`element attribute update attribute with false value to attribute with string value 1`] = `
<div
  width="10px"
/>
`;

exports[`element attribute update attribute with false value to attribute with true value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute update attribute with false value to attribute with undefined value 1`] = `<div />`;

exports[`element attribute update attribute with false value to empty attribute object 1`] = `<div />`;

exports[`element attribute update attribute with false value to undefined 1`] = `<div />`;

exports[`element attribute update attribute with string value to attribute with different value 1`] = `
<div
  width="20px"
/>
`;

exports[`element attribute update attribute with string value to attribute with empty string value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute update attribute with string value to attribute with false value 1`] = `<div />`;

exports[`element attribute update attribute with string value to attribute with same value 1`] = `
<div
  width="10px"
/>
`;

exports[`element attribute update attribute with string value to attribute with true value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute update attribute with string value to attribute with undefined value 1`] = `<div />`;

exports[`element attribute update attribute with string value to empty attribute object 1`] = `<div />`;

exports[`element attribute update attribute with string value to one attribute with same value and one with different 1`] = `
<div
  height="20px"
  width="10px"
/>
`;

exports[`element attribute update attribute with string value to two attributes with different values 1`] = `
<div
  height="20px"
  width="30px"
/>
`;

exports[`element attribute update attribute with string value to two different attributes 1`] = `
<div
  height="20px"
  left="30px"
/>
`;

exports[`element attribute update attribute with string value to undefined 1`] = `<div />`;

exports[`element attribute update attribute with true value to attribute with empty string value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute update attribute with true value to attribute with false value 1`] = `<div />`;

exports[`element attribute update attribute with true value to attribute with string value 1`] = `
<div
  width="10px"
/>
`;

exports[`element attribute update attribute with true value to attribute with true value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute update attribute with true value to attribute with undefined value 1`] = `<div />`;

exports[`element attribute update attribute with true value to empty attribute object 1`] = `<div />`;

exports[`element attribute update attribute with true value to undefined 1`] = `<div />`;

exports[`element attribute update attribute with undefined value to attribute with empty string value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute update attribute with undefined value to attribute with false value 1`] = `<div />`;

exports[`element attribute update attribute with undefined value to attribute with string value 1`] = `
<div
  width="10px"
/>
`;

exports[`element attribute update attribute with undefined value to attribute with true value 1`] = `
<div
  boolean=""
/>
`;

exports[`element attribute update attribute with undefined value to attribute with undefined value 1`] = `<div />`;

exports[`element attribute update attribute with undefined value to empty attribute object 1`] = `<div />`;

exports[`element attribute update attribute with undefined value to one attribute with same value and one with different 1`] = `
<div
  height="20px"
/>
`;

exports[`element attribute update attribute with undefined value to two attributes with different values 1`] = `
<div
  height="20px"
  width="30px"
/>
`;

exports[`element attribute update attribute with undefined value to two different attributes 1`] = `
<div
  height="20px"
  left="30px"
/>
`;

exports[`element attribute update attribute with undefined value to undefined 1`] = `<div />`;

exports[`element attribute update two attributes with string values to attribute with empty string value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute update two attributes with string values to attribute with undefined value 1`] = `<div />`;

exports[`element attribute update two attributes with string values to empty attribute object 1`] = `<div />`;

exports[`element attribute update two attributes with string values to one attribute with different value 1`] = `
<div
  width="30px"
/>
`;

exports[`element attribute update two attributes with string values to one attribute with same value and one with different 1`] = `
<div
  height="30px"
  width="10px"
/>
`;

exports[`element attribute update two attributes with string values to two attributes with different values 1`] = `
<div
  height="40px"
  width="30px"
/>
`;

exports[`element attribute update two attributes with string values to two attributes with same values 1`] = `
<div
  height="20px"
  width="10px"
/>
`;

exports[`element attribute update two attributes with string values to two different attributes 1`] = `
<div
  left="30px"
  right="40px"
/>
`;

exports[`element attribute update two attributes with string values to undefined 1`] = `<div />`;

exports[`element attribute update undefined to attribute with empty string value 1`] = `
<div
  width=""
/>
`;

exports[`element attribute update undefined to attribute with false value 1`] = `<div />`;

exports[`element attribute update undefined to attribute with string value 1`] = `
<div
  width="10px"
/>
`;

exports[`element attribute update undefined to attribute with true value 1`] = `
<div
  boolean=""
/>
`;

exports[`element attribute update undefined to attribute with undefined value 1`] = `<div />`;

exports[`element attribute update undefined to empty attributes object 1`] = `<div />`;

exports[`element attribute update undefined to two attributes with string values 1`] = `
<div
  height="20px"
  width="10px"
/>
`;

exports[`element attribute update undefined to undefined 1`] = `<div />`;
